### ISSUE TITLE
How to set the parameters while using laziness?

### ISSUE THREAD

**Issue by SparkJiao** [allenai/allennlp#1993]
I have added elmo to the bidaf.py but I find that it reported that the os can't allocate memory while I have modified the batch_size to 8 so I choose to use laziness, but the speed is so slow that it only trained two epochs where the batch size is 8 and the max instances in the memory is 1000, and the F1 is bad too. The F1 score can reach 65 without laziness while there is only 23.5 with elmo and after two epoch it get better a littile
```
2018-10-30 18:47:54,524 - INFO - allennlp.training.trainer - Training
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-30 18:47:55,322 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/c62aa48a5f15255e59d6f47f5362ed595b6e46de2450574980769e4ddcd381e0.bc13a91ce1137dd55c2bfcb244379148cd3279daa4f764570828ea3f0cb07cf3
2018-10-30 18:47:55,955 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.1773, end_acc: 0.1555, span_acc: 0.0565, em: 0.0176, f1: 0.1293, loss: 7.5337 ||: : 13581it [4:24:11,  1.09s/it]                
2018-10-30 23:12:06,758 - INFO - allennlp.training.trainer - Validating
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-30 23:12:07,910 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/3beb83a8f23ca2fac54061cac9317d5aee20c89801b6651ff4bade75a63cdf42.e0623be58f22c2be47092208906037002ac0eaa0052ab778f2cba4de8bb626a7
2018-10-30 23:12:08,024 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.3000, end_acc: 0.2841, span_acc: 0.1285, em: 0.0689, f1: 0.2634, loss: 5.7288 ||: : 998it [16:13,  1.79s/it]                   
2018-10-30 23:28:20,481 - INFO - allennlp.training.trainer -               Training |  Validation
2018-10-30 23:28:20,483 - INFO - allennlp.training.trainer - loss      |     7.534  |     5.729
2018-10-30 23:28:20,484 - INFO - allennlp.training.trainer - end_acc   |     0.156  |     0.284
2018-10-30 23:28:20,484 - INFO - allennlp.training.trainer - start_acc |     0.177  |     0.300
2018-10-30 23:28:20,485 - INFO - allennlp.training.trainer - span_acc  |     0.057  |     0.129
2018-10-30 23:28:20,485 - INFO - allennlp.training.trainer - em        |     0.018  |     0.069
2018-10-30 23:28:20,485 - INFO - allennlp.training.trainer - f1        |     0.129  |     0.263
2018-10-30 23:28:25,405 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to 'elmo_output/best.th'.
2018-10-30 23:28:25,774 - INFO - allennlp.training.trainer - Epoch duration: 04:40:31
2018-10-30 23:28:25,774 - INFO - allennlp.training.trainer - Estimated training time remaining: 3 days, 16:49:55
2018-10-30 23:28:25,774 - INFO - allennlp.training.trainer - Epoch 1/19
2018-10-30 23:28:25,774 - INFO - allennlp.training.trainer - Peak CPU memory usage MB: 5769.38
2018-10-30 23:28:25,907 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 7551
2018-10-30 23:28:25,908 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 10659
2018-10-30 23:28:25,908 - INFO - allennlp.training.trainer - GPU 2 memory usage MB: 10677
2018-10-30 23:28:25,908 - INFO - allennlp.training.trainer - GPU 3 memory usage MB: 11492
2018-10-30 23:28:25,909 - INFO - allennlp.training.trainer - Training
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-30 23:28:26,855 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/c62aa48a5f15255e59d6f47f5362ed595b6e46de2450574980769e4ddcd381e0.bc13a91ce1137dd55c2bfcb244379148cd3279daa4f764570828ea3f0cb07cf3
2018-10-30 23:28:30,073 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.2451, end_acc: 0.2272, span_acc: 0.0874, em: 0.0295, f1: 0.1860, loss: 6.5722 ||: : 13581it [4:31:13,  1.18s/it]               
2018-10-31 03:59:39,833 - INFO - allennlp.training.trainer - Validating
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-31 03:59:41,961 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/3beb83a8f23ca2fac54061cac9317d5aee20c89801b6651ff4bade75a63cdf42.e0623be58f22c2be47092208906037002ac0eaa0052ab778f2cba4de8bb626a7
2018-10-31 03:59:42,109 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.3084, end_acc: 0.2961, span_acc: 0.1315, em: 0.0858, f1: 0.2833, loss: 5.6663 ||: : 998it [16:21,  1.64s/it]                   
2018-10-31 04:16:02,084 - INFO - allennlp.training.trainer -               Training |  Validation
2018-10-31 04:16:02,085 - INFO - allennlp.training.trainer - loss      |     6.572  |     5.666
2018-10-31 04:16:02,086 - INFO - allennlp.training.trainer - end_acc   |     0.227  |     0.296
2018-10-31 04:16:02,086 - INFO - allennlp.training.trainer - start_acc |     0.245  |     0.308
2018-10-31 04:16:02,086 - INFO - allennlp.training.trainer - span_acc  |     0.087  |     0.132
2018-10-31 04:16:02,087 - INFO - allennlp.training.trainer - em        |     0.030  |     0.086
2018-10-31 04:16:02,087 - INFO - allennlp.training.trainer - f1        |     0.186  |     0.283
2018-10-31 04:16:13,146 - INFO - allennlp.training.trainer - Best validation performance so far. Copying weights to 'elmo_output/best.th'.
2018-10-31 04:16:13,604 - INFO - allennlp.training.trainer - Epoch duration: 04:47:47
2018-10-31 04:16:13,605 - INFO - allennlp.training.trainer - Estimated training time remaining: 3 days, 13:14:52
2018-10-31 04:16:13,605 - INFO - allennlp.training.trainer - Epoch 2/19
2018-10-31 04:16:13,612 - INFO - allennlp.training.trainer - Peak CPU memory usage MB: 5891.468
2018-10-31 04:16:15,575 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 7561
2018-10-31 04:16:15,575 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 10659
2018-10-31 04:16:15,575 - INFO - allennlp.training.trainer - GPU 2 memory usage MB: 10677
2018-10-31 04:16:15,576 - INFO - allennlp.training.trainer - GPU 3 memory usage MB: 11492
2018-10-31 04:16:15,577 - INFO - allennlp.training.trainer - Training
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-31 04:16:16,367 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/c62aa48a5f15255e59d6f47f5362ed595b6e46de2450574980769e4ddcd381e0.bc13a91ce1137dd55c2bfcb244379148cd3279daa4f764570828ea3f0cb07cf3
2018-10-31 04:16:17,748 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.2421, end_acc: 0.2221, span_acc: 0.0856, em: 0.0280, f1: 0.1842, loss: 6.7521 ||: : 13581it [4:24:24,  1.07s/it]               
2018-10-31 08:40:40,610 - INFO - allennlp.training.trainer - Validating
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-31 08:40:41,440 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/3beb83a8f23ca2fac54061cac9317d5aee20c89801b6651ff4bade75a63cdf42.e0623be58f22c2be47092208906037002ac0eaa0052ab778f2cba4de8bb626a7
2018-10-31 08:40:41,535 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
start_acc: 0.2940, end_acc: 0.2860, span_acc: 0.1184, em: 0.0639, f1: 0.2695, loss: 5.8808 ||: : 998it [16:07,  1.69s/it]                   
2018-10-31 08:56:48,947 - INFO - allennlp.training.trainer -               Training |  Validation
2018-10-31 08:56:48,949 - INFO - allennlp.training.trainer - loss      |     6.752  |     5.881
2018-10-31 08:56:48,949 - INFO - allennlp.training.trainer - end_acc   |     0.222  |     0.286
2018-10-31 08:56:48,950 - INFO - allennlp.training.trainer - start_acc |     0.242  |     0.294
2018-10-31 08:56:48,950 - INFO - allennlp.training.trainer - span_acc  |     0.086  |     0.118
2018-10-31 08:56:48,951 - INFO - allennlp.training.trainer - em        |     0.028  |     0.064
2018-10-31 08:56:48,951 - INFO - allennlp.training.trainer - f1        |     0.184  |     0.269
2018-10-31 08:57:05,570 - INFO - allennlp.training.trainer - Epoch duration: 04:40:51
2018-10-31 08:57:05,571 - INFO - allennlp.training.trainer - Estimated training time remaining: 3 days, 8:12:03
2018-10-31 08:57:05,571 - INFO - allennlp.training.trainer - Epoch 3/19
2018-10-31 08:57:05,571 - INFO - allennlp.training.trainer - Peak CPU memory usage MB: 5891.468
2018-10-31 08:57:05,732 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 12182
2018-10-31 08:57:05,732 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 10659
2018-10-31 08:57:05,733 - INFO - allennlp.training.trainer - GPU 2 memory usage MB: 10677
2018-10-31 08:57:05,733 - INFO - allennlp.training.trainer - GPU 3 memory usage MB: 11332
2018-10-31 08:57:05,734 - INFO - allennlp.training.trainer - Training
  0%|          | 0/1 [00:00<?, ?it/s]2018-10-31 08:57:06,525 - INFO - model.readers.bidaf_plus_dataset_reader - Reading file at /home/jiaofangkai/.allennlp/datasets/c62aa48a5f15255e59d6f47f5362ed595b6e46de2450574980769e4ddcd381e0.bc13a91ce1137dd55c2bfcb244379148cd3279daa4f764570828ea3f0cb07cf3
2018-10-31 08:57:07,187 - INFO - model.readers.bidaf_plus_dataset_reader - Reading the dataset
```
So I wanna know is there any way to solve this problem?

**Comment by SparkJiao** [comment]
Here is my config json file:
{
  "dataset_reader": {
    "type": "coqa-squad-plus",
    "lazy": true,
    "token_indexers": {
      "tokens": {
        "type": "single_id",
        "lowercase_tokens": true
      },
      "elmo": {
        "type": "elmo_characters"
      },
      "token_characters": {
        "type": "characters",
        "character_tokenizer": {
          "byte_encoding": "utf-8",
          "start_tokens": [
            259
          ],
          "end_tokens": [
            260
          ]
        }
      }
    }
  },
  "train_data_path": "https://nlp.stanford.edu/data/coqa/coqa-train-v1.0.json",
  "validation_data_path": "https://nlp.stanford.edu/data/coqa/coqa-dev-v1.0.json",
  "model": {
    "type": "bidaf",
    "text_field_embedder": {
      "token_embedders": {
        "tokens": {
          "type": "embedding",
          "pretrained_file": "https://s3-us-west-2.amazonaws.com/allennlp/datasets/glove/glove.6B.100d.txt.gz",
          "embedding_dim": 100,
          "trainable": false
        },
        "elmo": {
          "type": "elmo_token_embedder",
          "options_file": "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_options.json",
          "weight_file": "https://s3-us-west-2.amazonaws.com/allennlp/models/elmo/2x4096_512_2048cnn_2xhighway/elmo_2x4096_512_2048cnn_2xhighway_weights.hdf5",
          "do_layer_norm": false,
          "dropout": 0.5
        },
        "token_characters": {
          "type": "character_encoding",
          "embedding": {
            "num_embeddings": 262,
            "embedding_dim": 16
          },
          "encoder": {
            "type": "cnn",
            "embedding_dim": 16,
            "num_filters": 100,
            "ngram_filter_sizes": [
              5
            ]
          },
          "dropout": 0.2
        }
      }
    },
    "num_highway_layers": 2,
    "phrase_layer": {
      "type": "lstm",
      "bidirectional": true,
      "input_size": 1224,
      "hidden_size": 100,
      "num_layers": 1,
      "dropout": 0.2
    },
    "similarity_function": {
      "type": "linear",
      "combination": "x,y,x*y",
      "tensor_1_dim": 200,
      "tensor_2_dim": 200
    },
    "modeling_layer": {
      "type": "lstm",
      "bidirectional": true,
      "input_size": 800,
      "hidden_size": 100,
      "num_layers": 2,
      "dropout": 0.2
    },
    "span_end_encoder": {
      "type": "lstm",
      "bidirectional": true,
      "input_size": 1400,
      "hidden_size": 100,
      "num_layers": 1,
      "dropout": 0.2
    },
    "dropout": 0.2
  },
  "iterator": {
    "type": "bucket",
    "sorting_keys": [
      [
        "passage",
        "num_tokens"
      ],
      [
        "question",
        "num_tokens"
      ]
    ],
    "batch_size": 8,
    "max_instances_in_memory": 1000
  },
  "trainer": {
    "num_epochs": 20,
    "grad_norm": 5.0,
    "patience": 10,
    "validation_metric": "+em",
    "cuda_device": 3,
    "learning_rate_scheduler": {
      "type": "reduce_on_plateau",
      "factor": 0.5,
      "mode": "max",
      "patience": 2
    },
    "optimizer": {
      "type": "adam",
      "betas": [
        0.9,
        0.9
      ]
    }
  }
}
